discovery/discovery.go

```diff
@@ -32,6 +33,11 @@ import (
 //
 // Discoverers should initially send a full set of all discoverable TargetGroups.
 type Discoverer interface {
+	// A slice of prometheus collectors which need to be registered and unregistered:
+	// * Registration should happen prior to calling Run().
+	// * Unregistration should happen after Run() returns.
+	GetCollectors() []prometheus.Collector
```

Comment:
@beorn7 Do you think this new function is a good approach? If it looks good to you, I could implement it for all SD mechanisms.

If the manager handles registration, then it doesn't even need to pass a registerer to the SD. I personally think it would help reduce boilerplate and unintentional errors.

Another approach would be to have each SD manage its own registration and unregistration inside the `Run()` function, but I'm afraid this would be too manual and lead to errors.

Reply:
This would be an entirely new pattern (AFAIK). It feels weird to me that a `Discoverer` has to "hand out" internals, like the `prometheus.Collector` instances it owns, to the caller for some house keeping.

I would rather complete the `Discoverer` life cycle management by having it register all the Collectors on its own (as initially done in this PR) and then, before returning from the `Run` method, unregister them all again. In this way, we would also get rid of all the metrics related to an SD if that SD gets completely removed by a config change (rather than just restarted).

(Disclaimer: I'm not very familiar with the SD code. My idea above might not work.)

Reply:
> This would be an entirely new pattern (AFAIK). It feels weird to me that a `Discoverer` has to "hand out" internals, like the `prometheus.Collector` instances it owns, to the caller for some house keeping.

I agree, it does look weird.

> I would rather complete the `Discoverer` life cycle management by having it register all the Collectors on its own (as initially done in this PR) and then, before returning from the `Run` method, unregister them all again. In this way, we would also get rid of all the metrics related to an SD if that SD gets completely removed by a config change (rather than just restarted).

I'd be happy to let the SD manage their own registration, but would you mind if we do it a bit differently than before? Instead of registering the metrics as part of the [NewDiscoverer()](https://github.com/prometheus/prometheus/blob/v2.47.0/discovery/discovery.go#L57C2-L57C15), call, it would be better to pass a registerer via the [Run()](https://github.com/prometheus/prometheus/blob/v2.47.0/discovery/discovery.go#L38) function? This way we wouldn't "leak" a metric in case `NewDiscoverer` is called, but `Run` never gets called.

Reply:
This sounds good to me (given my limited insight ;) ). I'd say go ahead with that idea and see what breaks and if any of the devs with more SD experience intervene.

Reply:
So yes, I think the Registerer should be passed with the Run method, so that you don't need a separate Destroy method (as I have seen in the latest commit, but maybe that was added before you wrote above comment).

Let me know when I should commence reviewing.

Reply:
Apologies, I meant to write a GitHub comment last week about the `Destroy` method, but didn't get around to it. I think we have a choice between three approaches:

1. Register metrics when creating a new SD and unregister them by calling Destroy from the Manager:
   * This is what I did in the latest commit - [e30ce0f12758](https://github.com/prometheus/prometheus/pull/12565/commits/e30ce0f12758b95c3a7ac1ad64e27988d5ee12c1)
   * If a metric registration fails, it will fail during the creation of the SD, which to me seems more correct than failing during Run(). Metric registration feels like something which should be part of SD creation (the `NewDiscoverer()` method). Otherwise it would be possible to "create" multiple of the same SDs and only realise that they fail to run after calling Run on the ones after the 1st (since all but the 1st will fail to register their metric)
   * This means the SD Manager can unregister metrics by calling Destroy() even if the Run() function was never ran.
   * A downside to this approach is that it's easy for the Manager to miss calling a Destroy() in some situations. E.g. what if the creation of an SD failed, but metrics were still registered? Then Manager would still have to call Destroy().

2. Register metrics during Run() and unregistering them during Run()
   * The nice thing is that SD Manager doesn't have to call Destroy()
   * The bad thing is we only find out the registration failed after calling Run(). It feels a bit weird to me that the creation of an SD was successful, only to fail due to metric registration. 

3. Register metrics during creation of the SD, and unregister them during Run()
   * Again, the nice thing is Manager doesn't have to call a Destroy().
   * Run() should not be used for cleaning up resources required outside of Run() IMO. So I'd really prefer not to go with this approach.

Approach 1 seems the cleanest to me, but I'm happy to implement any approach that you think is reasonable. The big question is - which function do you think should fail if metric registration fails - `NewDiscoverer()` or `Run()`?

BTW I am planning to close this PR and instead will open separate PRs for different parts of the codebase - one PR for the scrape system, and another for the SDs. But it would be nice if we agree on an approach for the SDs prior to the new SD PR.

Reply:
I like (2) best. This is under the assumption that calling the `New...` function usually happens in close connection with calling the `Run` method, so that it actually doesn't make a big difference which method fails due to metrics collisions.

And even if calling `Run` happens much later after calling `New...`, there is the question what you will do if it fails. I assume in most cases you are probably going to exit the program execution anyway, which again means it doesn't make a big difference where the error is returned.